### PR TITLE
Add tests for the Bibtex parser

### PR DESCRIPTION
**Pull request by tobiasdiez**
I added a few tests for the Bibtex parser. I appreciate your feedback about the following points.
- [x] Please have a look at the `TODO` comments. These concern string concatenation and `@String` fields. For example in `parseRecognizesDateFieldWithConcatenation` the field `date = {1-4~} # nov` is parsed as `1-4~#nov#`. I wouldn't have expect the last sharp symbol.
- [x] Tests about JabRef's metadata (groups etc.) are still missing. Mostly, because I don't know the specifications :). Maybe in a new PR.
- [x] Related to the previous point, the `GUIGlobals.META_FLAG` flags should be moved to another place (they are not GUI, but parser/writer logic). What would be an appropriate new home for them?
- [x] There are 4 tests at the end of the file starting with `test`. These tests fail (without the `@Ignore` tag) but I'm not sure what they test and what would be the expected behavior in these cases.
- [x] There are quite a few ignored tests. These are related to open bugs in SF or are new bugs which I found while writing the tests (nothing big). I will open a new issue for the new bugs after this PR is merged.

For the record: While going through the code, I also changed the following things:
- Remove `BibtexParser.isRecognizedFormat` since the code has quite a few shortcomings (whitespace before @ or closing brackets on new lines were a problem) and the execution was quite slow for big files
- Fixed a bug where loading a file with corrupted bibtex entries showed an exception, cf. `parseIgnoresAndWarnsAboutCorruptedEntry()`
- Made `EntryType` an abstract class which implements the `Comparable` interface and overwrites `equals(Object other)`. Thus new expressions like `entry.getType() == BibtexEntryTypes.ARTICLE` return the expected result.

### CODE REVIEW

**Comment by koppor** [comment]
Regarding `#`: It might be that JabRef diverges from BibTeX's format. [BaseFrameHelp](http://jabref.sourceforge.net/help/BaseFrameHelp.php) states that `#` is used to reference strings. Is that BibTeX standard? [StringEditorHelp](http://jabref.sourceforge.net/help/StringEditorHelp.php) also states that "this syntax is specific for JabRef, and differs slightly from the bibtex notation that is produced when you save your database". Do you have time to investigate further?

Regarding your commits: The first two (a3a4b57 and 408a1a0) seem to be merge commits but also contain modifications by you. This is very hard to review. Proposal: make a *single* clean commit based on master.

1. (already done, just for the record) Ensure that `Jabref/master` is up to date. GitHub doc: https://help.github.com/articles/syncing-a-fork/
2. ensure that the latest `master` is merged into `parser`
3. `git checkout parser`
4. `git reset master`
5. commit - preferably using `git gui` to really investigate the differences.
6. commit and `git push -f`

Result: **one** commit in branch `parser` showing the difference to master.

**Comment by oscargus** [comment]
The sharp symbol is used differently in Bibtex and Jabref. In Jabref it is used to mark a string which shouldn't be enclosed in {} (or ""). In Bibtex it is used as a separator between enclosed text and strings. So I think it very much makes sense that `date = {1-4~} # nov` ends up as `1-4~#nov#` (which is what is shown in the JabRef entry editor, right?).

**Comment by oscargus** [comment]
To clarify further (if needed) `1-4~#nov#` in JabRef means that `1-4~` will be enclosed and `nov`is a string alias/key. In the Bibtex file it means `{1-4~}` or `"1-4~"` and `nov` concatenated with `#` inbetween.

**Comment by tobiasdiez** [comment]
@koppor I remember to have analyzed the regex used by `BibtexParser.isRecognizedFormat` and that I recognized some shortcomings (as stated above). While trying to fix them, I realized that there is no simple check to recognize a valid bibtex file without actually parsing it. And since the `isRecognizedFormat` actually tried to match a regex on every single line (in the worst case), I thought that it is an easier way to just parse every bib file and then state the errors. (Moreover, the method was used on import but not while opening a database). But if you prefer, I can reinsert it.

Yes, in bibtex the sharp symbol is used to concatenate strings as @oscargus explained. See also for example http://www.bibtex.org/Format/ . Apparently the current state works, so I just wanted to go safe and double-check if this is really the expected behavior. Thanks @oscargus for the clarifying words.

Ps: Yeah, I will clean up the git history...

**Comment by koppor** [comment]
Regarding the `#` issue: I would propose that we remove bibtex incompatible behavior. Is it possible that we exchange JabRef's character "`#`" to backtick "`"?

**Comment by koppor** [comment]
For me, it's OK to remove it. Just a comment in the code to this discussion. Maybe directly to your explanation above.

**Comment by oscargus** [comment]
I do not think it is incompatible, just confusing. :-) Using `#...#` to directly define strings in JabRef is quite well established and most of the time it is used "stand-alone", so only the string in the field (journal and month abbreviations). Somehow, to me, it only becomes a problem once things are concatenated and one tries to compare the bibtex field with the JabRef field and realize that the same character is used.

Backtick is anyway a bad choice since it is a combining accent (on most(?) keyboard layouts) and therefore require two key presses. (GitHub is really annoying me in that respect.)

**Comment by koppor** [comment]
The (German) keyboard layout [neo](http://www.neo-layout.org/) offers the backtick with a single key (and a modifier key).  :stuck_out_tongue_winking_eye:

I don't quite get the usage of `#...#`. Is it also written as such in the bib file or is it just a rendering thing in JabRef? I can live with the latter even though I find it confusing if the `bib` differs from the UI.

**Comment by oscargus** [comment]
Just rendering in JabRef. Try writing `#a#` in a field, it will show up as `a` (not `{a}`) in the bibtex source. Write `a#b#` and it will be `{a} # b`.

**Comment by tobiasdiez** [comment]
Could I please get feedback about the last two open points (`GUIGlobals.META_FLAG` and the tests about newlines). Thanks!

File: src/main/java/net/sf/jabref/importer/fileformat/BibtexImporter.java [hunk 1]
```diff
@@ -31,11 +31,11 @@
 public class BibtexImporter extends ImportFormat {
 
     /**
-     * @return false as that does not cause any harm in the current implementation of JabRef
+     * @return true as we have no effective way to decide whether a file is in bibtex format or not
      */
     @Override
     public boolean isRecognizedFormat(InputStream in) throws IOException {
-        return BibtexParser.isRecognizedFormat(new InputStreamReader(in));
+        return true;
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/importer/fileformat/BibtexImporter.java]: I do not like that we remove the logic to detect a bibtex file. On the contrary, we should extend this as we also want to detect whether a file is a bibtex or a biblatex file. This should not be removed.

File: src/main/java/net/sf/jabref/model/entry/EntryType.java [hunk 2]
```diff
@@ -52,10 +50,35 @@
...
+
+    @Override
+    public int hashCode() {
+        return this.getName().hashCode();
+    }
+
+    @Override
+    public int compareTo(EntryType o) {
+        return getName().compareTo(o.getName());
+    }
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/model/entry/EntryType.java]: I prefer interfaces over class extension. Would a Comparator<EntryType> also work for the use case here?

File: src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java [hunk 3]
```diff
@@ -32,219 +104,682 @@ public void testParseReader() throws IOException {
...
+        Assert.assertTrue(result.hasWarnings());
+
+        Collection<BibtexEntry> c = result.getDatabase().getEntries();
+        Assert.assertEquals(1, c.size());
+
+        BibtexEntry e = c.iterator().next();
+        Assert.assertEquals(BibtexEntryTypes.ARTICLE, e.getType());
+        Assert.assertEquals(null, e.getCiteKey());
+        Assert.assertEquals(2, e.getFieldNames().size());
+        Assert.assertEquals("Ed von Test", e.getField("author"));
```
Comment (simonharrer) [inline src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java]: Comparison should be done using the canonical bibtex format. Then, we can a) detect more errors, and b) the tests itself are getting easier to read.

**Comment by simonharrer** [comment]
> Made EntryType an abstract class which implements the Comparable interface and overwrites equals(Object other). Thus new expressions like entry.getType() == BibtexEntryTypes.ARTICLE return the expected result.

The second sentence is wrong. == never calls equals or hashCode or comparable, it simply checks if both reference the same instance in the heap. Because of this, I would not make the EntryType an abstract class. 

> Remove BibtexParser.isRecognizedFormat since the code has quite a few shortcomings (whitespace before @ or closing brackets on new lines were a problem) and the execution was quite slow for big files

To reiterate an inline comment: we aim to detect whether a file is a bibtex or biblatex file. This code can help with this. We should not delete it, but improve it. 

> Related to the previous point, the GUIGlobals.META_FLAG flags should be moved to another place (they are not GUI, but parser/writer logic). What would be an appropriate new home for them?

Agreed. Maybe a new class called JabRefDataFormat? It is currently used in MetaData and BibtexParser. I think we should assemble all read/write logic in a single place instead of distributing it over so many classes, as the MetaData class itself does not need to know how it can persist its structure. But I am open for other ideas regarding placing and naming. 

>  There are 4 tests at the end of the file starting with test. These tests fail (without the @Ignore tag) but I'm not sure what they test and what would be the expected behavior in these cases.

I think we should investigate. 2022983 refers to an issue number, somehow. Maybe to an old sourceforge number?

**Comment by koppor** [comment]
SourceForge changed global numbering of bugs to local numbering of bugs years ago. But they have redirects. Just use https://sourceforge.net/p/jabref/bugs/2022983/ as URL and it will redirect to https://sourceforge.net/p/jabref/bugs/679/.

File: src/main/java/net/sf/jabref/importer/fileformat/BibtexImporter.java (hunk 1 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/importer/fileformat/BibtexImporter.java]: The check equaled more or less to a search of the `@` symbol. So it was purely semantic. Moreover, there is difference between a Bibtex and a Biblatex file on this semantic level. As I have explained above, in my opinion there is no easy way to tell if a file is a valid bibtex file or not. This is why I have deleted this method (which, by the way was only called upon importing a bibtex file, but not on opening one).

File: src/main/java/net/sf/jabref/model/entry/EntryType.java (hunk 2 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/model/entry/EntryType.java]: I think `Comparator` is not applicable here. What is actually required is a proper equals method, the comparison is just an extra.

File: src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java (hunk 3 above)
Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java]: What is the canonical bibtex format? Do you mean, that the `BibtexEntry` should be converted to a bibtex string? I strongly oppose this change as then the parser tests could fail for two different reasons: the parser logic is wrong or the export to bibtex has a flaw.

Comment (simonharrer) [inline src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java]: Have a look at `CanonicalBibtexEntry`. This format will never change and can be used to compare bibtex entries with predefined texts.

**Comment by tobiasdiez** [comment]
Ok, I was wrong about the `==`. What I actually meant was that the line `Assert.assertEquals(BibtexEntryTypes.ARTICLE, e.getType());` works as expected. 

I liked the idea of a new `JabRefDataFormat` class and moved the meta flag there.

@koppor thanks for finding the old bug report! It says that the problem was multiple spaces in file names (which would explain the naming of the tests). However, I think this bug is not fixed: if you assign a file with multiple spaces to a entry. After saving and reopening the database, the file is no longer found. Somehow the spaces are lost during parsing, which is exactly why the tests fail.

File: src/main/java/net/sf/jabref/importer/fileformat/FieldContentParser.java [hunk 4]
```diff
@@ -37,7 +34,7 @@ public FieldContentParser() {
         multiLineFields.add("abstract");
         multiLineFields.add("review");
         // the file field should not be formatted, therefore we treat it as a multi line field
-        multiLineFields.addAll(Arrays.asList(Globals.prefs.getStringArray(JabRefPreferences.NON_WRAPPABLE_FIELDS)));
+        //multiLineFields.addAll(Arrays.asList(Globals.prefs.getStringArray(JabRefPreferences.NON_WRAPPABLE_FIELDS)));
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/importer/fileformat/FieldContentParser.java]: why did you comment out this line?

File: src/main/java/net/sf/jabref/importer/fileformat/BibtexParser.java [hunk 5]
```diff
@@ -283,7 +262,7 @@ public ParserResult parse() throws IOException {
                                     + be.getAuthorTitleYear(40) + " ("
                                     + Localization.lang("grouping may not work for this entry") + ")");
                         }
-                    } catch (IOException ex) {
+                    } catch (Exception ex) {
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/importer/fileformat/BibtexParser.java]: why did you change to Exception here? The code above does not state that Exception is required.

Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/importer/fileformat/BibtexParser.java]: If a bibtex field has for example no closing bracket, then another exception is thrown (not sure which concrete one, something along the lines of "I got something, but expected something else"). So I changed `IOException` to a general exception to handle also these cases. Without this change the test `parseIgnoresAndWarnsAboutCorruptedEntry` fails.

File: src/main/java/net/sf/jabref/importer/fileformat/FieldContentParser.java (hunk 4 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/importer/fileformat/FieldContentParser.java]: Not sure why. I'll revert it.

File: src/main/java/net/sf/jabref/importer/fileformat/BibtexParser.java (hunk 5 above)
Comment (lenhard) [inline src/main/java/net/sf/jabref/importer/fileformat/BibtexParser.java]: I think it would be better to find out the exact Exception (should be easy with the test) and then write a multi-catch block: `} catch (IOException | ThatOtherException ex) {`

If you catch `Exception`, you also catch Exceptions we do not want to catch, such as `NullPointerException`, `ArrayIndexOutOfBoundsException`, etc.

File: src/main/java/net/sf/jabref/importer/fileformat/FieldContentParser.java (hunk 4 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/importer/fileformat/FieldContentParser.java]: I remember (or better: the tests reminded me): This line was commented out because there were some null point exceptions due to uninitialized preferences. Fixed this in the new commit.

File: src/main/java/net/sf/jabref/importer/fileformat/BibtexParser.java (hunk 5 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/importer/fileformat/BibtexParser.java]: Yes you are right. I changed it to only catch `RuntimeExceptions`.

**Comment by tobiasdiez** [comment]
Ok, should I rebase (and thus probably delete some of your code comments)?

**Comment by simonharrer** [comment]
Can you please rebase this on master?

**Comment by tobiasdiez** [comment]
Rebased. 

It should be noted that I reversed a changed made in #469. Namely, Bibtexentry types are just compared by name (so a bibtex type with the same name as a biblatex type is considered the same). Otherwise all the tests would break in biblatex mode as then getType() returns the biblatex type.

I will propose a better solution tomorrow based on on the [state design pattern](https://en.wikipedia.org/wiki/State_pattern).

**Comment by tobiasdiez** [comment]
I removed all the code overwriting the `equals()` method for entry types. Thus now the PR just contains the added tests.From my side, this is ready to be merged in.

**Comment by simonharrer** [comment]
The issue with the isRecognizedFormat still persists. Will discuss in developer call how to proceed there. We need to check what effects will this have to set it to true. Maybe the current method is good enough?

**Comment by simonharrer** [comment]
Please rebase. Will merge this right away then. Dev call said true is OK for import format detection.

**Comment by tobiasdiez** [comment]
Rebased, but the findByDOI fails (probably a problem with the CI server not the code...)
